File: harmony-hdc/scripts/hdc_helpers.py
#!/usr/bin/env python3
"""Small CLI helpers for common HDC tasks on HarmonyOS."""

import argparse
import base64
import os
import re
import shlex
import subprocess
import sys
import time
from typing import Optional


def _hdc_prefix(serial: str | None) -> list[str]:
    if serial:
        return ["hdc", "-t", serial]
    return ["hdc"]


def _run(cmd: list[str], capture: bool = False, timeout: int = 15) -> subprocess.CompletedProcess:
    try:
        # print(f"DEBUG: running {' '.join(cmd)}")
        return subprocess.run(
            cmd, text=True, capture_output=capture, check=False, timeout=timeout
        )
    except subprocess.TimeoutExpired:
        return subprocess.CompletedProcess(
            args=cmd, returncode=124, stdout="", stderr="Command timed out"
        )


def cmd_devices(args: argparse.Namespace) -> int:
    """List connected devices."""
    result = _run(_hdc_prefix(args.serial) + ["list", "targets"], capture=True)
    if result.returncode != 0:
        print(result.stderr, file=sys.stderr)
        return result.returncode

    # HDC list output is just identifiers, sometimes with status
    # We just print it out, but cleanup empty lines
    for line in result.stdout.split("\n"):
        if line.strip():
            print(line.strip())
    return 0


def cmd_connect(args: argparse.Namespace) -> int:
    """Connect to remote device."""
    addr = args.address
    if ":" not in addr:
        addr = f"{addr}:5555"
    result = _run(["hdc", "tconn", addr], capture=False)
    return result.returncode


def cmd_disconnect(args: argparse.Namespace) -> int:
    """Disconnect remote device."""
    cmd = ["hdc", "tdisconn"]
    if args.address:
        cmd.append(args.address)
    # else: hdc doesnt support bulk disconnect easily without scripting
    result = _run(cmd, capture=False)
    return result.returncode


def cmd_get_ip(args: argparse.Namespace) -> int:
    """Get device IP address via ifconfig."""
    cmd = _hdc_prefix(args.serial) + ["shell", "ifconfig"]
    result = _run(cmd, capture=True)

    for line in result.stdout.split("\n"):
        if "inet addr:" in line or "inet " in line:
            parts = line.strip().split()
            for i, part in enumerate(parts):
                if "addr:" in part:
                    ip = part.split(":")[1]
                    if not ip.startswith("127."):
                        print(ip)
                        return 0
                elif part == "inet" and i + 1 < len(parts):
                    ip = parts[i + 1].split("/")[0]
                    if not ip.startswith("127."):
                        print(ip)
                        return 0

    print("IP not found", file=sys.stderr)
    return 1


def cmd_shell(args: argparse.Namespace) -> int:
    """Run raw shell command."""
    cmd = _hdc_prefix(args.serial) + ["shell"] + args.command
    result = _run(cmd, capture=False)
    return result.returncode


def cmd_tap(args: argparse.Namespace) -> int:
    """Tap at coordinates using uitest."""
    cmd = _hdc_prefix(args.serial) + ["shell", "uitest", "uiInput", "click", str(args.x), str(args.y)]
    result = _run(cmd, capture=False)
    return result.returncode


def cmd_double_tap(args: argparse.Namespace) -> int:
    """Double tap at coordinates using uitest."""
    cmd = _hdc_prefix(args.serial) + ["shell", "uitest", "uiInput", "doubleClick", str(args.x), str(args.y)]
    result = _run(cmd, capture=False)
    return result.returncode


def cmd_swipe(args: argparse.Namespace) -> int:
    """Swipe using uitest."""
    cmd = _hdc_prefix(args.serial) + [
        "shell",
        "uitest",
        "uiInput",
        "swipe",
        str(args.x1),
        str(args.y1),
        str(args.x2),
        str(args.y2),
    ]
    if args.duration_ms is not None:
        cmd.append(str(args.duration_ms))
    result = _run(cmd, capture=False)
    return result.returncode


def cmd_keyevent(args: argparse.Namespace) -> int:
    """Send key event."""
    # HarmonyOS uses key codes or names (e.g. Back, Home)
    cmd = _hdc_prefix(args.serial) + ["shell", "uitest", "uiInput", "keyEvent", str(args.keycode)]
    result = _run(cmd, capture=False)
    return result.returncode


def cmd_text(args: argparse.Namespace) -> int:
    """Input text."""
    # Escape quotes and special chars
    escaped = args.text.replace('"', '\\"').replace("$", "\\$")

    # hdc shell uitest uiInput text "content"
    cmd = _hdc_prefix(args.serial) + ["shell", "uitest", "uiInput", "text", escaped]
    result = _run(cmd, capture=False)

    # Auto-enter if multi-line (simulated by caller logic usually, but here simple version)
    return result.returncode


def cmd_screenshot(args: argparse.Namespace) -> int:
    """Capture screenshot."""
    out_path = args.out or "screen.png"
    remote_path = "/data/local/tmp/tmp_screenshot.jpeg"
    prefix = _hdc_prefix(args.serial)

    # Try 'screenshot' command
    res = _run(prefix + ["shell", "screenshot", remote_path], capture=True)
    if "fail" in res.stdout.lower() or "error" in res.stdout.lower() or "not found" in res.stdout.lower():
         # Fallback to snapshot_display
         _run(prefix + ["shell", "snapshot_display", "-f", remote_path], capture=True)

    # Pull
    res_pull = _run(prefix + ["file", "recv", remote_path, out_path], capture=False)

    # Cleanup
    _run(prefix + ["shell", "rm", remote_path], capture=False)

    return res_pull.returncode


def cmd_launch(args: argparse.Namespace) -> int:
    """Launch app (Ability) by bundle info."""
    # User passes 'bundle' or 'bundle/ability'.
    # If no ability, we assume 'EntryAbility' (common default) or user should provide it.

    bundle = args.package
    ability = "EntryAbility"

    if "/" in bundle:
        bundle, ability = bundle.split("/", 1)

    cmd = _hdc_prefix(args.serial) + [
        "shell",
        "aa",
        "start",
        "-b",
        bundle,
        "-a",
        ability
    ]
    result = _run(cmd, capture=False)
    return result.returncode


def cmd_force_stop(args: argparse.Namespace) -> int:
    """Force stop app."""
    cmd = _hdc_prefix(args.serial) + ["shell", "aa", "force-stop", args.package]
    result = _run(cmd, capture=False)
    return result.returncode


def cmd_get_current_app(args: argparse.Namespace) -> int:
    """Get current bundle name from aa dump."""
    cmd = _hdc_prefix(args.serial) + ["shell", "aa", "dump", "-l"]
    result = _run(cmd, capture=True)
    output = result.stdout

    lines = output.split("\n")
    foreground_bundle = None
    current_bundle = None

    for line in lines:
        if "app name [" in line:
            match = re.search(r'\[([^\]]+)\]', line)
            if match:
                current_bundle = match.group(1)

        if "state #FOREGROUND" in line or "state #foreground" in line.lower():
            if current_bundle:
                foreground_bundle = current_bundle
                break

        if "Mission ID" in line:
            current_bundle = None

    if foreground_bundle:
        print(foreground_bundle)
        return 0
    else:
        print("System Home", file=sys.stderr)
        return 1


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="HDC helper commands for HarmonyOS")
    parser.add_argument("-s", "--serial", "-t", dest="serial", help="device serial/id (matches -t in hdc)")

    sub = parser.add_subparsers(dest="command", required=True)

    sub.add_parser("devices", help="list devices").set_defaults(func=cmd_devices)

    p_conn = sub.add_parser("connect", help="connect remote device")
    p_conn.add_argument("address")
    p_conn.set_defaults(func=cmd_connect)

    p_disc = sub.add_parser("disconnect", help="disconnect remote device")
    p_disc.add_argument("address", nargs="?")
    p_disc.set_defaults(func=cmd_disconnect)

    sub.add_parser("get-ip", help="get device ip").set_defaults(func=cmd_get_ip)

    p_shell = sub.add_parser("shell", help="run hdc shell command")
    p_shell.add_argument("command", nargs=argparse.REMAINDER)
    p_shell.set_defaults(func=cmd_shell)

    p_tap = sub.add_parser("tap", help="tap x y")
    p_tap.add_argument("x", type=int)
    p_tap.add_argument("y", type=int)
    p_tap.set_defaults(func=cmd_tap)

    p_dtap = sub.add_parser("double-tap", help="double tap x y")
    p_dtap.add_argument("x", type=int)
    p_dtap.add_argument("y", type=int)
    p_dtap.set_defaults(func=cmd_double_tap)

    p_swipe = sub.add_parser("swipe", help="swipe x1 y1 x2 y2")
    p_swipe.add_argument("x1", type=int)
    p_swipe.add_argument("y1", type=int)
    p_swipe.add_argument("x2", type=int)
    p_swipe.add_argument("y2", type=int)
    p_swipe.add_argument("--duration-ms", type=int, help="duration in ms (200-5000)")
    p_swipe.set_defaults(func=cmd_swipe)

    p_key = sub.add_parser("keyevent", help="send key code/name (e.g. Back, Home)")
    p_key.add_argument("keycode")
    p_key.set_defaults(func=cmd_keyevent)

    p_text = sub.add_parser("text", help="input text")
    p_text.add_argument("text")
    p_text.set_defaults(func=cmd_text)

    p_shot = sub.add_parser("screenshot", help="capture screenshot")
    p_shot.add_argument("--out", help="output path")
    p_shot.set_defaults(func=cmd_screenshot)

    p_launch = sub.add_parser("launch", help="launch app (bundle or bundle/ability)")
    p_launch.add_argument("package", help="bundle name or bundle/AbilityName")
    p_launch.set_defaults(func=cmd_launch)

    p_stop = sub.add_parser("force-stop", help="force stop app")
    p_stop.add_argument("package")
    p_stop.set_defaults(func=cmd_force_stop)

    sub.add_parser("get-current-app", help="get foreground bundle").set_defaults(func=cmd_get_current_app)

    return parser


def main() -> int:
    parser = build_parser()
    args = parser.parse_args()
    return args.func(args)


if __name__ == "__main__":
    raise SystemExit(main())

File: harmony-hdc/scripts/hdc_helpers.ts
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";
import chalk from "chalk";

const defaultTimeoutMs = 15_000;

type LogLevel = "debug" | "info" | "error";

function createLogger(json: boolean, level: LogLevel, stream: NodeJS.WriteStream, color: boolean) {
  const levels: Record<LogLevel, number> = { debug: 10, info: 20, error: 40 };
  const min = levels[level] ?? 20;
  const useColor = color && !json;
  const levelColor = (lv: LogLevel, value: string) => {
    if (!useColor) return value;
    return lv === "error" ? chalk.red(value) : lv === "debug" ? chalk.cyan(value) : chalk.green(value);
  };
  const keyColor = (value: string) => (useColor ? chalk.blue(value) : value);
  const msgColor = (value: string) => (useColor ? chalk.green(value) : value);
  const valueColor = (value: string) => (useColor ? chalk.dim(value) : value);
  function shouldLog(lv: LogLevel) {
    return levels[lv] >= min;
  }
  function formatValue(value: unknown): string {
    if (typeof value === "string") {
      if (value === "") return '""';
      if (/\s/.test(value) || value.includes("=") || value.includes("\"")) {
        return JSON.stringify(value);
      }
      return value;
    }
    if (value === null || value === undefined) return "null";
    if (typeof value === "number" || typeof value === "boolean") return String(value);
    return JSON.stringify(value);
  }
  function write(lv: LogLevel, msg: string, fields?: Record<string, unknown>) {
    if (!shouldLog(lv)) return;
    const time = new Date().toISOString();
    if (json) {
      const payload: Record<string, unknown> = { time, level: lv.toUpperCase(), msg };
      if (fields) for (const [k, v] of Object.entries(fields)) payload[k] = v;
      stream.write(`${JSON.stringify(payload)}\n`);
      return;
    }
    const parts = [
      `${keyColor("time")}=${valueColor(time)}`,
      `${keyColor("level")}=${levelColor(lv, lv.toUpperCase())}`,
      `${keyColor("msg")}=${msgColor(msg)}`,
    ];
    if (fields) {
      for (const [k, v] of Object.entries(fields)) {
        parts.push(`${keyColor(k)}=${valueColor(formatValue(v))}`);
      }
    }
    stream.write(parts.join(" ") + "\n");
  }
  return {
    debug: (msg: string, fields?: Record<string, unknown>) => write("debug", msg, fields),
    info: (msg: string, fields?: Record<string, unknown>) => write("info", msg, fields),
    error: (msg: string, fields?: Record<string, unknown>) => write("error", msg, fields),
  };
}

let logger = createLogger(false, "info", process.stderr, Boolean(process.stderr.isTTY));

function hdcPrefix(serial: string) {
  return serial ? ["hdc", "-t", serial] : ["hdc"];
}

type CmdResult = { stdout: string; stderr: string; exitCode: number };

function runCmd(cmd: string[], capture: boolean, timeoutMs: number): CmdResult {
  const res = spawnSync(cmd[0], cmd.slice(1), {
    encoding: "utf8",
    timeout: timeoutMs > 0 ? timeoutMs : defaultTimeoutMs,
    stdio: capture ? "pipe" : "inherit",
  });
  if (res.error) {
    if ((res.error as any).code === "ETIMEDOUT") {
      return { stdout: res.stdout ?? "", stderr: "Command timed out", exitCode: 124 };
    }
    return { stdout: res.stdout ?? "", stderr: res.stderr ?? String(res.error), exitCode: 1 };
  }
  const exitCode = typeof res.status === "number" ? res.status : 0;
  return { stdout: res.stdout ?? "", stderr: res.stderr ?? "", exitCode };
}

function runHdc(serial: string, capture: boolean, args: string[]) {
  return runCmd([...hdcPrefix(serial), ...args], capture, defaultTimeoutMs);
}

function cmdDevices(serial: string) {
  const result = runHdc(serial, true, ["list", "targets"]);
  if (result.exitCode !== 0) {
    if (result.stderr) process.stderr.write(result.stderr.trim() + "\n");
    return result.exitCode;
  }
  for (const line of result.stdout.split("\n")) {
    if (line.trim()) console.log(line.trim());
  }
  return 0;
}

function cmdConnect(args: string[]) {
  if (args.length < 1) {
    process.stderr.write("connect requires <address>\n");
    return 2;
  }
  let addr = args[0];
  if (!addr.includes(":")) addr += ":5555";
  return runCmd(["hdc", "tconn", addr], false, defaultTimeoutMs).exitCode;
}

function cmdDisconnect(args: string[]) {
  const cmd = ["hdc", "tdisconn", ...(args[0] ? [args[0]] : [])];
  return runCmd(cmd, false, defaultTimeoutMs).exitCode;
}

function cmdGetIP(serial: string) {
  const result = runHdc(serial, true, ["shell", "ifconfig"]);
  for (let line of result.stdout.split("\n")) {
    line = line.trim();
    if (!line) continue;
    if (line.includes("inet addr:")) {
      const parts = line.split(/\s+/);
      for (const part of parts) {
        if (part.startsWith("addr:")) {
          const ip = part.replace("addr:", "");
          if (ip && !ip.startsWith("127.")) {
            console.log(ip);
            return 0;
          }
        }
      }
    }
    if (line.includes("inet ")) {
      const parts = line.split(/\s+/);
      for (let i = 0; i < parts.length; i++) {
        if (parts[i] === "inet" && i + 1 < parts.length) {
          const ip = parts[i + 1].split("/")[0];
          if (ip && !ip.startsWith("127.")) {
            console.log(ip);
            return 0;
          }
        }
      }
    }
  }
  process.stderr.write("IP not found\n");
  return 1;
}

function cmdShell(serial: string, args: string[]) {
  return runHdc(serial, false, ["shell", ...args]).exitCode;
}

function cmdTap(serial: string, args: string[]) {
  if (args.length < 2) {
    process.stderr.write("tap requires <x> <y>\n");
    return 2;
  }
  return runHdc(serial, false, ["shell", "uitest", "uiInput", "click", args[0], args[1]]).exitCode;
}

function cmdDoubleTap(serial: string, args: string[]) {
  if (args.length < 2) {
    process.stderr.write("double-tap requires <x> <y>\n");
    return 2;
  }
  return runHdc(serial, false, ["shell", "uitest", "uiInput", "doubleClick", args[0], args[1]]).exitCode;
}

function cmdSwipe(serial: string, args: string[], durationMs: number) {
  if (args.length < 4) {
    process.stderr.write("swipe requires <x1> <y1> <x2> <y2>\n");
    return 2;
  }
  const cmd = ["shell", "uitest", "uiInput", "swipe", args[0], args[1], args[2], args[3]];
  if (durationMs >= 0) cmd.push(String(durationMs));
  return runHdc(serial, false, cmd).exitCode;
}

function cmdKeyEvent(serial: string, args: string[]) {
  if (args.length < 1) {
    process.stderr.write("keyevent requires <keycode>\n");
    return 2;
  }
  return runHdc(serial, false, ["shell", "uitest", "uiInput", "keyEvent", args[0]]).exitCode;
}

function cmdText(serial: string, args: string[]) {
  if (args.length < 1) {
    process.stderr.write("text requires <text>\n");
    return 2;
  }
  let text = args.join(" ");
  text = text.replace(/"/g, "\\\"").replace(/\$/g, "\\$");
  return runHdc(serial, false, ["shell", "uitest", "uiInput", "text", text]).exitCode;
}

function cmdScreenshot(serial: string, outPath: string) {
  let target = outPath || "screen.png";
  try {
    target = path.resolve(target);
  } catch {
    // ignore
  }
  const remotePath = "/data/local/tmp/tmp_screenshot.jpeg";
  const prefix = hdcPrefix(serial);
  const res = runCmd([...prefix, "shell", "screenshot", remotePath], true, defaultTimeoutMs);
  const resOut = res.stdout.toLowerCase();
  if (resOut.includes("fail") || resOut.includes("error") || resOut.includes("not found")) {
    runCmd([...prefix, "shell", "snapshot_display", "-f", remotePath], true, defaultTimeoutMs);
  }
  const pull = runCmd([...prefix, "file", "recv", remotePath, target], false, defaultTimeoutMs);
  runCmd([...prefix, "shell", "rm", remotePath], false, defaultTimeoutMs);
  return pull.exitCode;
}

function cmdLaunch(serial: string, args: string[]) {
  if (args.length < 1) {
    process.stderr.write("launch requires <bundle> or <bundle/ability>\n");
    return 2;
  }
  let bundle = args[0];
  let ability = "EntryAbility";
  if (bundle.includes("/")) {
    const parts = bundle.split("/", 2);
    bundle = parts[0];
    ability = parts[1];
  }
  return runHdc(serial, false, ["shell", "aa", "start", "-b", bundle, "-a", ability]).exitCode;
}

function cmdForceStop(serial: string, args: string[]) {
  if (args.length < 1) {
    process.stderr.write("force-stop requires <bundle>\n");
    return 2;
  }
  return runHdc(serial, false, ["shell", "aa", "force-stop", args[0]]).exitCode;
}

function cmdGetCurrentApp(serial: string) {
  const result = runHdc(serial, true, ["shell", "aa", "dump", "-l"]);
  const output = result.stdout;
  const lines = output.split("\n");
  let foreground = "";
  let current = "";
  const re = /\[([^\]]+)\]/;
  for (const line of lines) {
    if (line.includes("app name [")) {
      const match = line.match(re);
      if (match && match[1]) current = match[1];
    }
    if (line.includes("state #FOREGROUND") || line.toLowerCase().includes("state #foreground")) {
      if (current) {
        foreground = current;
        break;
      }
    }
    if (line.includes("Mission ID")) current = "";
  }
  if (foreground) {
    console.log(foreground);
    return 0;
  }
  process.stderr.write("System Home\n");
  return 1;
}

function hasHelpArg(args: string[]) {
  return args.some((arg) => arg === "-h" || arg === "--help");
}

function printUsage(out: NodeJS.WriteStream) {
  out.write("Usage:\n");
  out.write("  hdc_helpers [flags] <command> [args]\n\n");
  out.write("Commands:\n");
  out.write("  devices\n");
  out.write("  connect <address>\n");
  out.write("  disconnect [address]\n");
  out.write("  get-ip\n");
  out.write("  shell <cmd...>\n");
  out.write("  tap <x> <y>\n");
  out.write("  double-tap <x> <y>\n");
  out.write("  swipe <x1> <y1> <x2> <y2> [--duration-ms N]\n");
  out.write("  keyevent <keycode>\n");
  out.write("  text <text>\n");
  out.write("  screenshot [--out path]\n");
  out.write("  launch <bundle>[/Ability]\n");
  out.write("  force-stop <bundle>\n");
  out.write("  get-current-app\n\n");
  out.write("Global Flags:\n");
  out.write("  -t, -s <id>    device serial/id\n");
}

function parseGlobalFlags(args: string[]) {
  let serial = "";
  const rest: string[] = [];
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === "-t" || arg === "-s") {
      if (i + 1 < args.length) {
        serial = args[i + 1];
        i++;
      }
      continue;
    }
    if (arg.startsWith("-t=")) {
      serial = arg.split("=")[1] ?? "";
      continue;
    }
    rest.push(arg);
  }
  return { serial, rest };
}

function parseCommandFlags(args: string[]) {
  const flags: Record<string, string | boolean> = {};
  const rest: string[] = [];
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === "-h" || arg === "--help") {
      flags.help = true;
      continue;
    }
    if (arg === "--duration-ms") {
      i++;
      if (i >= args.length) return { error: "--duration-ms requires value" } as const;
      flags["duration-ms"] = args[i];
      continue;
    }
    if (arg.startsWith("--duration-ms=")) {
      flags["duration-ms"] = arg.split("=")[1] ?? "";
      continue;
    }
    if (arg === "--out") {
      i++;
      if (i >= args.length) return { error: "--out requires value" } as const;
      flags.out = args[i];
      continue;
    }
    if (arg.startsWith("--out=")) {
      flags.out = arg.split("=")[1] ?? "";
      continue;
    }
    if (!arg.startsWith("-") || arg === "-") {
      rest.push(...args.slice(i));
      break;
    }
    return { error: `unknown flag ${arg}` } as const;
  }
  return { flags, rest } as const;
}

async function main() {
  const argv = process.argv.slice(2);
  const global = parseGlobalFlags(argv);
  const args = global.rest;
  if (args.length === 0 || args[0] === "help" || args[0] === "-h" || args[0] === "--help") {
    printUsage(process.stdout);
    process.exit(0);
  }
  const cmd = args[0];
  const cmdArgs = args.slice(1);
  switch (cmd) {
    case "devices":
      process.exit(cmdDevices(global.serial));
    case "connect":
      process.exit(cmdConnect(cmdArgs));
    case "disconnect":
      process.exit(cmdDisconnect(cmdArgs));
    case "get-ip":
      process.exit(cmdGetIP(global.serial));
    case "shell":
      process.exit(cmdShell(global.serial, cmdArgs));
    case "tap":
      process.exit(cmdTap(global.serial, cmdArgs));
    case "double-tap":
      process.exit(cmdDoubleTap(global.serial, cmdArgs));
    case "swipe": {
      const parsed = parseCommandFlags(cmdArgs);
      if ("error" in parsed) {
        process.stderr.write(`${parsed.error}\n`);
        process.exit(2);
      }
      if (parsed.flags?.help) {
        process.stdout.write("Usage:\n  hdc_helpers swipe [flags] <x1> <y1> <x2> <y2>\n\n");
        process.stdout.write("Flags:\n  --duration-ms <ms>    swipe duration in ms\n");
        process.exit(0);
      }
      const duration = parsed.flags?.["duration-ms"] ? Number(parsed.flags["duration-ms"]) : -1;
      process.exit(cmdSwipe(global.serial, parsed.rest ?? [], Number.isFinite(duration) ? duration : -1));
    }
    case "keyevent":
      process.exit(cmdKeyEvent(global.serial, cmdArgs));
    case "text":
      process.exit(cmdText(global.serial, cmdArgs));
    case "screenshot": {
      const parsed = parseCommandFlags(cmdArgs);
      if ("error" in parsed) {
        process.stderr.write(`${parsed.error}\n`);
        process.exit(2);
      }
      if (parsed.flags?.help) {
        process.stdout.write("Usage:\n  hdc_helpers screenshot [flags]\n\n");
        process.stdout.write("Flags:\n  --out <path>    output path\n");
        process.exit(0);
      }
      process.exit(cmdScreenshot(global.serial, String(parsed.flags?.out || "")));
    }
    case "launch":
      process.exit(cmdLaunch(global.serial, cmdArgs));
    case "force-stop":
      process.exit(cmdForceStop(global.serial, cmdArgs));
    case "get-current-app":
      process.exit(cmdGetCurrentApp(global.serial));
    default:
      process.stderr.write(`Unknown command: ${cmd}\n`);
      printUsage(process.stdout);
      process.exit(2);
  }
}

main().catch((err) => {
  logger.error("unhandled error", { err: err?.message || String(err) });
  process.exit(1);
});
